File: src/subside/plugins/koth/commands/CommandMode.java
package subside.plugins.koth.commands;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import org.bukkit.command.CommandSender;

import subside.plugins.koth.captureentities.CaptureTypeRegistry;
import subside.plugins.koth.commands.CommandHandler.CommandCategory;
import subside.plugins.koth.exceptions.CommandMessageException;
import subside.plugins.koth.gamemodes.GamemodeRegistry;
import subside.plugins.koth.modules.Lang;
import subside.plugins.koth.utils.IPerm;
import subside.plugins.koth.utils.MessageBuilder;
import subside.plugins.koth.utils.Perm;

public class CommandMode extends AbstractCommand {

    public CommandMode(CommandCategory category) {
        super(category);
    }

    @Override
    public void run(CommandSender sender, String[] args) {
        if(args.length > 0){
            String[] newArgs = Arrays.copyOfRange(args, 1, args.length);
            if(args[0].equalsIgnoreCase("gamemode")){
                gameMode(sender, newArgs);
                return;
            } else if(args[0].equalsIgnoreCase("captureType")) {
                captureType(sender, newArgs);
                return;
            }
        }
        
        List<String> list = new ArrayList<>();
        list.addAll(new MessageBuilder(Lang.COMMAND_GLOBAL_HELP_TITLE).title("KoTH mode commands").buildArray());
        list.addAll(new MessageBuilder(Lang.COMMAND_GLOBAL_HELP_INFO).command("/koth mode capturetype").commandInfo("Switch between capute types").buildArray());
        list.addAll(new MessageBuilder(Lang.COMMAND_GLOBAL_HELP_INFO).command("/koth mode gamemode").commandInfo("Switch between gamemodes").buildArray());
        sender.sendMessage(list.toArray(new String[list.size()]));
    }
    
    public void captureType(CommandSender sender, String[] args){
        CaptureTypeRegistry cER = getPlugin().getCaptureTypeRegistry();
        if(args.length > 0){
            if(cER.getCaptureTypes().containsKey(args[0].toLowerCase())){
                cER.setPreferedClass(cER.getCaptureTypes().get(args[0].toLowerCase()));

                throw new CommandMessageException(new MessageBuilder(Lang.COMMAND_ENTITY_CHANGED).entry(args[0]));
            } else {
                throw new CommandMessageException(new MessageBuilder(Lang.COMMAND_ENTITY_NOT_EXIST).entry(args[0]));
            }
        } else {
            new MessageBuilder(Lang.COMMAND_ENTITY_LIST_TITLE).buildAndSend(sender);

            for (String capEntity : cER.getCaptureTypes().keySet()) {
                new MessageBuilder(Lang.COMMAND_ENTITY_LIST_ENTRY).entry(capEntity).buildAndSend(sender);
            }
        }
    }

    public void gameMode(CommandSender sender, String[] args) {
        GamemodeRegistry gR = getPlugin().getGamemodeRegistry();
        if (args.length > 0) {
            if (gR.getGamemodes().containsKey(args[0].toLowerCase())) {
                gR.setCurrentMode(args[0]);
                throw new CommandMessageException(new MessageBuilder(Lang.COMMAND_MODE_CHANGED).entry(args[0]));
            } else {
                throw new CommandMessageException(new MessageBuilder(Lang.COMMAND_MODE_NOT_EXIST).entry(args[0]));
            }
        } else {
            new MessageBuilder(Lang.COMMAND_MODE_LIST_TITLE).buildAndSend(sender);

            for (String gamemode : gR.getGamemodes().keySet()) {
                new MessageBuilder(Lang.COMMAND_MODE_LIST_ENTRY).entry(gamemode).buildAndSend(sender);
            }
        }
    }

    @Override
    public IPerm getPermission() {
        return Perm.Admin.MODE;
    }

    @Override
    public String[] getCommands() {
        return new String[] {
                "mode"
        };
    }
    
    @Override
    public String getUsage() {
        return "/koth mode [mode]";
    }

    @Override
    public String getDescription() {
        return "change the gamemode";
    }

}

File: src/subside/plugins/koth/commands/control/CommandMode.kt
package subside.plugins.koth.commands.control

import org.bukkit.command.CommandSender
import subside.plugins.koth.KothManager
import subside.plugins.koth.commands.Command
import subside.plugins.koth.modules.Lang
import subside.plugins.koth.utils.IPerm
import subside.plugins.koth.utils.MessageBuilder
import subside.plugins.koth.utils.Perm
import java.util.*

class CommandMode : Command {
    override fun run(kothManager: KothManager, sender: CommandSender, args: Array<String>) {
        if (args.isEmpty()) {
            showHelp(sender)
            return
        }

        val newArgs = offsetArgs(args, 1)
        when(args[0].toLowerCase()) {
            "gamemode" -> gameMode(kothManager, sender, newArgs)
            "captureType" -> captureType(kothManager, sender, newArgs)
            else -> showHelp(sender)
        }
    }

    private fun captureType(kothManager: KothManager, sender: CommandSender, args: Array<String>) {
        if (args.isEmpty()) {
            MessageBuilder(Lang.COMMAND_ENTITY_LIST_TITLE).send(sender)

            kothManager.captureTypeRegistry.captureTypes.forEach { (key, _) ->
                MessageBuilder(Lang.COMMAND_ENTITY_LIST_ENTRY).entry(key).send(sender)
            }
            return
        }

        val captureType = kothManager.captureTypeRegistry.captureTypes.get(args[0])
        if (captureType == null) {
            MessageBuilder(Lang.COMMAND_ENTITY_NOT_EXIST)
                    .entry(args[0])
                    .send(sender)
            return
        }

        kothManager.captureTypeRegistry.preferedClass = captureType
        MessageBuilder(Lang.COMMAND_ENTITY_CHANGED)
                .entry(args[0])
                .send(sender)
    }

    private fun gameMode(kothManager: KothManager, sender: CommandSender, args: Array<String>) {
        if (args.isEmpty()) {
            MessageBuilder(Lang.COMMAND_MODE_LIST_TITLE).send(sender)

            kothManager.gamemodeRegistry.gamemodes.forEach { (key, _) ->
                MessageBuilder(Lang.COMMAND_MODE_LIST_ENTRY).entry(key).send(sender)
            }
            return
        }

        val gamemode = kothManager.gamemodeRegistry.gamemodes.get(args[0].toLowerCase())
        if (gamemode == null) {
            MessageBuilder(Lang.COMMAND_MODE_NOT_EXIST)
                    .entry(args[0])
                    .send(sender)
            return
        }

        kothManager.gamemodeRegistry.currentMode = gamemode
        MessageBuilder(Lang.COMMAND_MODE_CHANGED)
                .entry(args[0])
                .send(sender)
    }

    private fun showHelp(sender: CommandSender) {
        val list = ArrayList<String>()
        list.addAll(MessageBuilder(Lang.COMMAND_GLOBAL_HELP_TITLE).title("KoTH mode commands").buildArray())
        list.addAll(MessageBuilder(Lang.COMMAND_GLOBAL_HELP_INFO).command("/koth mode capturetype").commandInfo("Switch between capute types").buildArray())
        list.addAll(MessageBuilder(Lang.COMMAND_GLOBAL_HELP_INFO).command("/koth mode gamemode").commandInfo("Switch between gamemodes").buildArray())
        sender.sendMessage(list.toTypedArray())
    }

    override val permission: IPerm = Perm.Admin.MODE
    override val commands: Array<String> = arrayOf("mode")
    override val usage: String = "/koth mode [mode]"
    override val description: String = "change the gamemode"
}